airbyte-integrations/connectors/destination-keen/src/main/java/io/airbyte/integrations/destination/keen/KeenTimestampService.java

```diff
@@ -0,0 +1,187 @@
+package io.airbyte.integrations.destination.keen;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.joestelmach.natty.Parser;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KeenTimestampService {
+
+  public enum CursorType {STRING, NUMBER, UNRECOGNIZED}
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KeenRecordsConsumer.class);
+
+  private static final long SECONDS_FROM_EPOCH_THRESHOLD = 1_000_000_000L;
+
+  private static final long MILLIS_FROM_EPOCH_THRESHOLD = 10_000_000_000L;
+
+  // Map containing stream names paired with their cursor fields
+  private Map<String, CursorField> streamCursorFields;
+  private final Parser parser;
+  private final boolean timestampInferenceEnabled;
+
+  public KeenTimestampService(ConfiguredAirbyteCatalog catalog, boolean timestampInferenceEnabled) {
+    this.streamCursorFields = new HashMap<>();
+    this.parser = new Parser();
+    this.timestampInferenceEnabled = timestampInferenceEnabled;
+
+    if (timestampInferenceEnabled) {
+      LOGGER.info("Initializing KeenTimestampService, finding cursor fields.");
+      streamCursorFields = catalog.getStreams()
+          .stream()
+          .filter(stream -> stream.getCursorField().size() == 1)
+          .map(s -> Pair.of(s.getStream().getName(), CursorField.fromStream(s)))
+          .filter(
+              pair -> pair.getRight() != null && pair.getRight().type != CursorType.UNRECOGNIZED)
+          .collect(Collectors.toMap(Pair::getKey, Pair::getValue));
+    }
+  }
+
+  /**
+   * Tries to inject keen.timestamp field to the given message data. If the stream contains
+   * cursor fields of types NUMBER or STRING, this value is tried to be parsed to a timestamp.
+   * If this procedure fails, stream is removed from timestamp-parsable stream map, so parsing is not tried
+   * for future messages in the same stream. If parsing succeeds, keen.timestamp field is put as a JSON node
+   * to the message data and whole data is returned. Otherwise, keen.timestamp is set to emittedAt value
+   * @param message AirbyteMessage containing record data
+   * @return Record data together with keen.timestamp field
+   */
+  public JsonNode injectTimestamp(AirbyteMessage message) {
```

Comment:
we should accept `AirbyteRecordMessage` instead just for type safety

Reply:
fixed